### PR TITLE
tests: verify timestamp recovery

### PR DESCRIPTION
**Pull request by lukaszstolarczuk**
- [x] wait for #201

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)

### CODE REVIEW

**Review by lukaszstolarczuk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)* status: 0 of 17 files reviewed, 1 unresolved discussion

___
*[`tests/integrity/append_timestamp_recovery.gdb` line 22 at r1](https://reviewable.io/reviews/pmem/pmemstream/192#-N2rJrVt-nx4Ly3rSMDR:-N2vt4McCMJOP_BbjuP1:b-xdger1) ([raw file](https://github.com/pmem/pmemstream/blob/923713ce3c88d42f0cb1a645e80547c79475a47a/tests/integrity/append_timestamp_recovery.gdb#L22)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

when I've torn the append in a different place (this line uncommented; not only with `+600`, but also `+10`) gdb breaks (because it does not fail at `UT_ASSERT_UNREACHABLE;`), but in "check_consistency" phase it shows a count of 4 elements in the region...
</blockquote></details>

at this moment I think there might be a bug in the recovery. I'm pretty sure the break is happening at the memcpy part (in pmemstream_publish) and I still have 4 entries read from the region (in the 'integrity check' part of the test.

... to be continued :wink:

**Review by igchor — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)* status: 0 of 17 files reviewed, 1 unresolved discussion (waiting on @igchor and @lukaszstolarczuk)

___
*[`tests/integrity/append_timestamp_recovery.gdb` line 22 at r1](https://reviewable.io/reviews/pmem/pmemstream/192#-N2rJrVt-nx4Ly3rSMDR:-N2w6XhN2tJgO_O8IHDV:bt91l3q) ([raw file](https://github.com/pmem/pmemstream/blob/923713ce3c88d42f0cb1a645e80547c79475a47a/tests/integrity/append_timestamp_recovery.gdb#L22)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

at this moment I think there might be a bug in the recovery. I'm pretty sure the break is happening at the memcpy part (in pmemstream_publish) and I still have 4 entries read from the region (in the 'integrity check' part of the test.

... to be continued :wink: 
</blockquote></details>

Don't we already have a very similar test which is actually passing? Few things to check:
- is the entry actually fully written to pmem?
- what is the timestamp of the last entry

Also, to actually check the region recovery, you should write the test slightly differently.

You should break AFTER the timestamp is actually written to PMEM, but before we updated persisted_timestamp. If there is a crash during memcpy, no recovery should be needed (no timestamp is written yet).

Also, on integrity check you should append the entry to DIFFERENT region. This way we would have two entires with the same timestamp and we can see if this is handled properly.

**Review by lukaszstolarczuk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)* status: 0 of 17 files reviewed, 1 unresolved discussion (waiting on @igchor)

___
*[`tests/integrity/append_timestamp_recovery.gdb` line 22 at r1](https://reviewable.io/reviews/pmem/pmemstream/192#-N2rJrVt-nx4Ly3rSMDR:-N2wzUb-BQcqjRd3gXe2:b-qg8dub) ([raw file](https://github.com/pmem/pmemstream/blob/923713ce3c88d42f0cb1a645e80547c79475a47a/tests/integrity/append_timestamp_recovery.gdb#L22)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Don't we already have a very similar test which is actually passing? Few things to check:
- is the entry actually fully written to pmem?
- what is the timestamp of the last entry

Also, to actually check the region recovery, you should write the test slightly differently.

You should break AFTER the timestamp is actually written to PMEM, but before we updated persisted_timestamp. If there is a crash during memcpy, no recovery should be needed (no timestamp is written yet).

Also, on integrity check you should append the entry to DIFFERENT region. This way we would have two entires with the same timestamp and we can see if this is handled properly.
</blockquote></details>

if you're talking about `append_break` - then yeah, it's similar, it's actually a (temporary) copy of that entire test - the main difference is that we break in publish (not in reserve part) and we check for timestamp after break.

Right now, as you can see in the log of this test:
- we break during publish's memcpy (so the entry is not fully written and the timestamp is acquired)
- we check integrity and count 4 elements (`assertion failure: cnt (0x4) == init_data.size() (0x3)`)

don't we expect here only 3 elements?

// I'll use the rest of your tips, when I figure out this question...

**Review by igchor — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)* status: 0 of 17 files reviewed, 1 unresolved discussion (waiting on @igchor and @lukaszstolarczuk)

___
*[`tests/integrity/append_timestamp_recovery.gdb` line 22 at r1](https://reviewable.io/reviews/pmem/pmemstream/192#-N2rJrVt-nx4Ly3rSMDR:-N2x_lQ6DPN8nxQ8fQ8h:br8048h) ([raw file](https://github.com/pmem/pmemstream/blob/923713ce3c88d42f0cb1a645e80547c79475a47a/tests/integrity/append_timestamp_recovery.gdb#L22)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

if you're talking about `append_break` - then yeah, it's similar, it's actually a (temporary) copy of that entire test - the main difference is that we break in publish (not in reserve part) and we check for timestamp after break.

Right now, as you can see in the log of this test:
- we break during publish's memcpy (so the entry is not fully written and the timestamp is acquired)
- we check integrity and count 4 elements (`assertion failure: cnt (0x4) == init_data.size() (0x3)`)

don't we expect here only 3 elements?

// I'll use the rest of your tips, when I figure out this question...
</blockquote></details>

Ok, I see, can you print what is the value of max_valid_timestamp for the region?
It should be equal to persisted_offset, which should be 3 I belive (so only entries with timestamps 0,1,2 would be valid).

**Review by lukaszstolarczuk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)* status: 0 of 18 files reviewed, 1 unresolved discussion (waiting on @igchor and @lukaszstolarczuk)

___
*[`tests/integrity/append_timestamp_recovery.gdb` line 22 at r1](https://reviewable.io/reviews/pmem/pmemstream/192#-N2rJrVt-nx4Ly3rSMDR:-N3-QXWj5KGZiqFALbeI:b-obtg93) ([raw file](https://github.com/pmem/pmemstream/blob/923713ce3c88d42f0cb1a645e80547c79475a47a/tests/integrity/append_timestamp_recovery.gdb#L22)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

Ok, I see, can you print what is the value of max_valid_timestamp for the region?
It should be equal to persisted_offset, which should be 3 I belive (so only entries with timestamps 0,1,2 would be valid).
</blockquote></details>

actually, the problem was I torn the memcpy after the first 8bits, so the timestamp wasn't set properly.
I'm thinking that case is probably not the real case scenario (correct me if I'm wrong).

Nonetheless, I believe this led me to the problem and my proposal of the solution - pls see https://github.com/pmem/pmemstream/pull/196 (the same code is applied here, so the test passes).

**Review by lukaszstolarczuk — COMMENTED** [review]
@igchor, I've ended up adding 2 new "breaks", each testing a bit different tearing of an entry.

> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192#-:-N420Eub79rhfipWjeBW:b-u9e60u)* status: 0 of 23 files reviewed, all discussions resolved (waiting on @igchor)

**Review by lukaszstolarczuk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)* status: 2 of 25 files reviewed, 2 unresolved discussions (waiting on @igchor)

___
*[`tests/integrity/append_break.cpp` line 53 at r6](https://reviewable.io/reviews/pmem/pmemstream/192#-N46TOel2eTX65YrLUnv:-N4RkANo6pO_QAtJoWEr:bobo4ov) ([raw file](https://github.com/pmem/pmemstream/blob/484b15f563ae7cbbc0b650f1cc14f75789938a00/tests/integrity/append_break.cpp#L53)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

We should also have a test which appends to the second (empty region) but does not iterate over the elements prior to append. This is because our recovery is kind of "lazy".

This kind of test would actually result in having two entires physically written to stream with the same timestamps (one in first region and one in second region). Only after we recover the first region (.e.g during iteration AFTER append) we will invalidate that first entry.

You can even use span_helpers to verify that we actually have 2 entries with the same timestamp (and not causing any recovery).
</blockquote></details>

partially done (just pushed a current version to see if my thinking is alright)
___
*[`tests/integrity/append_break_1.gdb` line 17 at r6](https://reviewable.io/reviews/pmem/pmemstream/192#-N46Sx1K4OXMWRjvo2AM:-N4RfkXP1jUu_0Dt6aNA:b-896fix) ([raw file](https://github.com/pmem/pmemstream/blob/484b15f563ae7cbbc0b650f1cc14f75789938a00/tests/integrity/append_break_1.gdb#L17)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

can we verify it somehow? (We probably could if we had logs, maybe add some TODO here?)
</blockquote></details>

Done.

**Review by lukaszstolarczuk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)* status: 2 of 25 files reviewed, 2 unresolved discussions (waiting on @igchor)

___
*[`tests/integrity/append_break.cpp` line 53 at r6](https://reviewable.io/reviews/pmem/pmemstream/192#-N46TOel2eTX65YrLUnv:-N4WQ1fQDuPtixMgv9UG:b-896fix) ([raw file](https://github.com/pmem/pmemstream/blob/484b15f563ae7cbbc0b650f1cc14f75789938a00/tests/integrity/append_break.cpp#L53)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

partially done (just pushed a current version to see if my thinking is alright)
</blockquote></details>

Done.

**Review by igchor — APPROVED** [review]
:lgtm:

> Reviewed 2 of 5 files at r5, 5 of 5 files at r7.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192#-:-N4aSoof8NHNoX8a8yGz:bnfp4nl)* status: 8 of 25 files reviewed, 2 unresolved discussions (waiting on @igchor and @lukaszstolarczuk)

___
*[`tests/integrity/append_break.cpp` line 118 at r7](https://reviewable.io/reviews/pmem/pmemstream/192#-N4aS2t0D_TQdMbSW1U-:-N4aS2t0D_TQdMbSW1U0:b6oho83) ([raw file](https://github.com/pmem/pmemstream/blob/e4e1ccbee3aba266fb64a2d3b5b898067af6b0d6/tests/integrity/append_break.cpp#L118)):*
> ```cpp
> 		/* check if we have for sure a duplicated timestamp */
> 		auto regions = span_runtimes_from_stream(s.sut, 0, UINT64_MAX);
> 		UT_ASSERTeq(regions[0].sub_spans.size(), 5); /* 4 entry spans (incl. broken one) + 1 empty span */
> ```

I'm not sure if this is always true. Will we always have this empty span at the end? I think it will only work if the entire stream is zeroed-out at the beginning - do we do that each time?

Maybe it would be better to just check regions[0].sub_spans.size() >= 4 and that we actually have 4 entry spans?
___
*[`tests/integrity/append_break.cpp` line 122 at r7](https://reviewable.io/reviews/pmem/pmemstream/192#-N4aSTS68WjZaTTheEPq:-N4aSTS68WjZaTTheEPr:b-x23e3q) ([raw file](https://github.com/pmem/pmemstream/blob/e4e1ccbee3aba266fb64a2d3b5b898067af6b0d6/tests/integrity/append_break.cpp#L122)):*
> ```cpp
> 										       regions[0].sub_spans[3].offset);
> 		UT_ASSERT(span_get_type(&broken_entry_in_r1->span_base) == SPAN_ENTRY);
> 		UT_ASSERTeq(regions[1].sub_spans.size(), 2); /* 1 entry + 1 empty span */
> ```

Here it's OK, since append completed and we surely zeroed out next entry.

**Review by lukaszstolarczuk — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)* status: 7 of 25 files reviewed, all discussions resolved (waiting on @igchor)

___
*[`tests/integrity/append_break.cpp` line 118 at r7](https://reviewable.io/reviews/pmem/pmemstream/192#-N4aS2t0D_TQdMbSW1U-:-N4aUOqq8vWYOury2GT5:b-jhxt30) ([raw file](https://github.com/pmem/pmemstream/blob/e4e1ccbee3aba266fb64a2d3b5b898067af6b0d6/tests/integrity/append_break.cpp#L118)):*
<details><summary><i>Previously, igchor (Igor Chorążewicz) wrote…</i></summary><blockquote>

I'm not sure if this is always true. Will we always have this empty span at the end? I think it will only work if the entire stream is zeroed-out at the beginning - do we do that each time?

Maybe it would be better to just check regions[0].sub_spans.size() >= 4 and that we actually have 4 entry spans?
</blockquote></details>

I think we handle that when creating a region (and in append), but we might change that in the future, so, to be safe, I think it's good to update this.

**Review by igchor — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192#-:-N4aZaOZ4BEc0-K7NTx9:bnfp4nl)* status: 7 of 25 files reviewed, all discussions resolved (waiting on @igchor)

**Review by KFilipek — COMMENTED** [review]
> Reviewed 1 of 17 files at r1, 13 of 16 files at r2, 5 of 11 files at r4, 1 of 1 files at r8, 1 of 1 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @lukaszstolarczuk)

File: tests/integrity/append_break.cpp [hunk 1]
```diff
@@ -57,7 +57,7 @@ static void test(char mode)
 
 int main(int argc, char *argv[])
 {
-	if (argc != 3 || strchr("abi", argv[1][0]) == nullptr) {
+	if (argc != 3 || strchr("abiy", argv[1][0]) == nullptr) {
```
Comment (karczex) [inline tests/integrity/append_break.cpp:187]: How this line corresponds with commit: "extend cmake helpers with separate executable param "

**Review by karczex — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/pmem/pmemstream/192)* status: all files reviewed, 2 unresolved discussions (waiting on @lukaszstolarczuk)

___
*[`tests/integrity/append_break.cpp` line 96 at r9](https://reviewable.io/reviews/pmem/pmemstream/192#-N4ayTt3D2tektB-98Ud:-N4ayTt3D2tektB-98Ue:b-1fkyh9) ([raw file](https://github.com/pmem/pmemstream/blob/94f35d6fcc4e1beb8ed3be4e1a1037bb8792e355/tests/integrity/append_break.cpp#L96)):*
> ```cpp
> 		UT_ASSERT(read_elements.back() == buf);
> 
> 	} else if (mode == 'y') {
> ```

Maybe change modes names to human readable form - like it's done in `singly_linked_list_pmredored.cpp'
___
*[`tests/integrity/append_break.cpp` line 188 at r9](https://reviewable.io/reviews/pmem/pmemstream/192#-N4ayGmw1YY8WC1p7R8O:-N4ayGmw1YY8WC1p7R8P:bv6nhk5) ([raw file](https://github.com/pmem/pmemstream/blob/94f35d6fcc4e1beb8ed3be4e1a1037bb8792e355/tests/integrity/append_break.cpp#L188)):*
> ```cpp
> {
> 	if (argc != 3 || strchr("abiy", argv[1][0]) == nullptr) {
> 		std::cout << "Usage: " << argv[0] << " <a|b|i> file-path" << std::endl;
> ```

You forgot `y` in help message.
